File: examples/payment-reconciliation/01-naive-ingest/sim.py
"""The bug: a naive webhook handler with no dedup and no state machine.
Run this, then 02-idempotent-ingest, to see what fixing ingest alone does
and doesn't repair. Run with: python3 sim.py"""

import random
from dataclasses import dataclass, field

SUCCEEDED = "charge.succeeded"
REFUNDED = "charge.refunded"
N = 10_000


@dataclass(frozen=True)
class Event:
    id: str
    charge_id: str
    type: str


@dataclass(frozen=True)
class Charge:
    id: str
    rail_state: str  # ground truth at the rail


@dataclass(frozen=True)
class Channel:
    loss: float
    dup: float
    reorder: float


def generate(n, rng):
    return [
        Charge(f"ch_{i:05d}", "refunded" if rng.random() < 0.10 else "succeeded")
        for i in range(n)
    ]


def true_events(c):
    ev = [Event(f"{c.id}_e1", c.id, SUCCEEDED)]
    if c.rail_state == "refunded":
        ev.append(Event(f"{c.id}_e2", c.id, REFUNDED))
    return ev


def deliver(ev, ch, rng):
    out = list(ev)
    if len(out) > 1 and rng.random() < ch.reorder:
        i = rng.randrange(len(out) - 1)
        out[i], out[i + 1] = out[i + 1], out[i]

    delivered = []
    for e in out:
        if rng.random() < ch.loss:
            continue
        delivered.append(e)
        if rng.random() < ch.dup:
            delivered.append(e)
    return delivered


@dataclass
class DB:
    state: dict = field(default_factory=dict)
    credits: dict = field(default_factory=dict)

    def naive(self, e):
        """No dedup, no state machine. Last write wins; side effect every time."""
        if e.type == SUCCEEDED:
            self.state[e.charge_id] = "succeeded"
            self.credits[e.charge_id] = self.credits.get(e.charge_id, 0) + 1
        elif e.type == REFUNDED:
            self.state[e.charge_id] = "refunded"


def measure(db, charges):
    stuck = double = wrong = 0
    for c in charges:
        mine = db.state.get(c.id, "unknown")
        if db.credits.get(c.id, 0) > 1:
            double += 1
        if mine in ("unknown", "pending"):
            stuck += 1
        elif mine != c.rail_state:
            wrong += 1
    return stuck, double, wrong


if __name__ == "__main__":
    ch = Channel(loss=0.05, dup=0.10, reorder=0.05)
    rng = random.Random(42)
    charges = generate(N, rng)

    db = DB()
    for c in charges:
        for e in deliver(true_events(c), ch, rng):
            db.naive(e)

    stuck, double, wrong = measure(db, charges)

    def pct(x):
        return f"{100 * x / N:.2f}%"

    print(f"naive ingest — {N} charges, loss=5% dup=10% reorder=5%, seed=42")
    print(f"  stuck in pending : {pct(stuck)}   (rail moved money, we never noticed)")
    print(f"  double-credited  : {pct(double)}   (duplicate webhook, side effect fired twice)")
    print(f"  wrong terminal   : {pct(wrong)}   (out-of-order webhook flipped the final state)")
    print()
    print("This is the bug: no dedup, no order tolerance, no idea anything is wrong.")

File: examples/payment-reconciliation/internal/simcore/sim.go
// Package simcore is the shared engine behind the three runnable programs in
// this directory (01-naive-ingest, 02-idempotent-ingest, 03-reconciliation).
// It models one payment rail, one unreliable webhook channel, and the ingest
// strategies built in the article. All three programs seed the same PRNG in
// the same order, so they replay the identical delivery transcript — any
// difference in their printed results comes from the ingest logic alone.
package simcore

import (
	"fmt"
	"math/rand"
)

type Event struct {
	ID       string
	ChargeID string
	Type     string // "charge.succeeded" | "charge.refunded"
}

type Charge struct {
	ID        string
	RailState string // ground truth at the rail
}

// Channel is what the network does to webhooks on the way to us.
type Channel struct {
	Loss    float64 // webhook never arrives
	Dup     float64 // rail retried; we receive it twice
	Reorder float64 // events arrive in the wrong order
}

func Generate(n int, rng *rand.Rand) []Charge {
	charges := make([]Charge, n)
	for i := range charges {
		state := "succeeded"
		if rng.Float64() < 0.10 {
			state = "refunded" // 10% of charges are later refunded
		}
		charges[i] = Charge{ID: fmt.Sprintf("ch_%05d", i), RailState: state}
	}
	return charges
}

// TrueEvents is what actually happened at the rail, in order.
func TrueEvents(c Charge) []Event {
	ev := []Event{{ID: c.ID + "_e1", ChargeID: c.ID, Type: "charge.succeeded"}}
	if c.RailState == "refunded" {
		ev = append(ev, Event{ID: c.ID + "_e2", ChargeID: c.ID, Type: "charge.refunded"})
	}
	return ev
}

// Deliver applies the channel's failure modes to one charge's event stream.
func Deliver(ev []Event, ch Channel, rng *rand.Rand) []Event {
	out := append([]Event(nil), ev...)

	if len(out) > 1 && rng.Float64() < ch.Reorder {
		i := rng.Intn(len(out) - 1)
		out[i], out[i+1] = out[i+1], out[i]
	}

	delivered := make([]Event, 0, len(out)+1)
	for _, e := range out {
		if rng.Float64() < ch.Loss {
			continue
		}
		delivered = append(delivered, e)
		if rng.Float64() < ch.Dup {
			delivered = append(delivered, e)
		}
	}
	return delivered
}

type DB struct {
	state   map[string]string
	credits map[string]int // the side effect duplicates corrupt
	claimed map[string]bool
}

func NewDB() *DB {
	return &DB{
		state:   map[string]string{},
		credits: map[string]int{},
		claimed: map[string]bool{},
	}
}

// Naive: no dedup, no state machine. Last write wins; side effect every time.
// This is Version 1 in the article.
func (db *DB) Naive(e Event) {
	switch e.Type {
	case "charge.succeeded":
		db.state[e.ChargeID] = "succeeded"
		db.credits[e.ChargeID]++
	case "charge.refunded":
		db.state[e.ChargeID] = "refunded"
	}
}

// Hardened: claim the event id, then a state machine that tolerates disorder.
// This is Version 2 in the article.
func (db *DB) Hardened(e Event) {
	if db.claimed[e.ID] {
		return
	}
	db.claimed[e.ID] = true

	cur, ok := db.state[e.ChargeID]
	if !ok {
		cur = "unknown"
	}
	switch e.Type {
	case "charge.succeeded":
		if cur == "unknown" || cur == "pending" {
			db.state[e.ChargeID] = "succeeded"
			db.credits[e.ChargeID]++
		}
	case "charge.refunded":
		db.state[e.ChargeID] = "refunded" // terminal: accept from any state
	}
}

type ReconResult struct{ Repaired, Alerted int }

// Reconcile is Version 3 in the article: diff into buckets, auto-repair only
// the lost-webhook case, alert on everything else.
func Reconcile(db *DB, charges []Charge) ReconResult {
	var r ReconResult
	for _, c := range charges {
		mine, ok := db.state[c.ID]
		if !ok {
			mine = "unknown"
		}
		switch {
		case mine == c.RailState:
			// matched: the boring, good case
		case mine == "unknown" || mine == "pending":
			// only-in-rail: the lost-webhook case, safe to auto-repair
			db.Hardened(Event{
				ID:       "recon-" + c.ID,
				ChargeID: c.ID,
				Type:     "charge." + c.RailState,
			})
			r.Repaired++
		default:
			// both terminal and disagreeing: never auto-repair
			r.Alerted++
		}
	}
	return r
}

type Metrics struct {
	StuckPending  int
	DoubleCredit  int
	WrongTerminal int
}

func (m Metrics) Disagree() int { return m.StuckPending + m.WrongTerminal }

func Measure(db *DB, charges []Charge) Metrics {
	var m Metrics
	for _, c := range charges {
		mine, ok := db.state[c.ID]
		if !ok {
			mine = "unknown"
		}
		if db.credits[c.ID] > 1 {
			m.DoubleCredit++
		}
		if mine == "unknown" || mine == "pending" {
			m.StuckPending++
		} else if mine != c.RailState {
			m.WrongTerminal++
		}
	}
	return m
}

func Pct(x, n int) string { return fmt.Sprintf("%.2f%%", 100*float64(x)/float64(n)) }
